### ISSUE TITLE
How to not overfit on timeseries problems?

### ISSUE THREAD

**Issue by jakobhuss** [perpetual-ml/perpetual#78]
Hi! 

I tried this model on timeseries. I might be incorrect but i got the feeling that the model kept training due to leakage between samples. Perhaps due to date time features.

Usually i keep a val split that is perhaps two weeks of data after the training set.

time --->
| -------------- train ------------ |----- val ---- | ----- test ----- |

Is perpetual suitable for these types of problems?

**Comment by deadsoul44** [comment]
PerpetualBooster is suitable for time series, but data should be stationary. In other words, you should use differentiated target and features that are time related because generalization control uses a cross validation like logic to control how much trees should grow.

**Comment by jakobhuss** [comment]
I'm aware that the time series should be stationary. I think I get leakage during training causing the training to hit iteration limits. Usually I need some kind of mechanism similar to scikit TimeSeriesSplit instead of classic cv-splits to make sure i don't use the future to predict the past. The problem is known as "temporal leakage".

Looking at the api i might be able to use the prune() method on my validation set. But that would probably be very inefficient.

I hope I'm not missing anything and wasting your time.
Thank you

**Comment by jakobhuss** [comment]
I tried an experiment that i think for me indicates perpetual over fits due to temporal leakage. To note is that perpetual takes more than 10x time than LGBM. I also included a LGBM with early stopping on validation set which performs the best.

I don't know the inner details of the algorithm but for time series i think you need to be able to specify what samples to validate on to avoid over fitting.

```python
class TimeSeriesLGBM(LGBMRegressor):
    def __init__(self, ts_split_ratio=0.98, early_stop_rounds=50, **kwargs):
        self.ts_split_ratio = ts_split_ratio
        self.early_stop_rounds = early_stop_rounds
        super().__init__(**kwargs)

    def fit(self, X, y, **fit_params):
        # 1. Chronological split
        split = int(len(X) * self.ts_split_ratio)
        X_train, X_val = X[:split], X[split:]
        y_train, y_val = y[:split], y[split:]

        # 2. Add early stopping to callbacks
        callbacks = fit_params.pop("callbacks", [])
        callbacks.append(early_stopping(self.early_stop_rounds, verbose=False))

        return super().fit(
            X_train, y_train, eval_set=[(X_val, y_val)], callbacks=callbacks, **fit_params
        )


lgbm_default = LGBMRegressor(objective="mse")

lgbm_valid = TimeSeriesLGBM(objective="mse", max_iter=1000)

perpetual = PerpetualBooster(objective="SquaredLoss", budget=1.0)
```

```txt
metric                           MASE   MAE      RMSE
split model                     
train lgbm-default               1.9772 193.0441 258.4297
      lgbm-with-validation-stop  1.7837 174.1514 236.5104
      perpetual                  1.4487 141.4441 225.3814


test  lgbm-default               3.5394 481.2557 682.2049
      lgbm-with-validation-stop  3.5051 476.5938 670.8077
      perpetual                  3.8163 518.9099 707.5186
```

**Comment by deadsoul44** [comment]
Is this a univariate or multivariate dataset? Is there any chance to share the dataset? Otherwise, I will test with an open source dataset.
